lib/vm/src/vmcontext.rs

```diff
@@ -15,15 +15,48 @@ use std::ptr::{self, NonNull};
 use std::sync::Arc;
 use std::u32;
 
+/// Union representing the first parameter passed when calling a function.
+///
+/// It may either be a pointer to the [`VMContext`] if it's a Wasm function
+/// or a pointer to arbitrary data controlled by the host if it's a host function.
+#[derive(Copy, Clone)]
+pub union VMFunctionExtraData {
+    /// Wasm functions take a pointer to [`VMContext`].
+    pub vmctx: *mut VMContext,
+    /// Host functions can have custom environments.
+    pub host_env: *mut std::ffi::c_void,
```

Comment:
As far as I understand this PR (which backport and enhance (!) a pre-1.0 feature), `host_env` can hold 2 kind of pointers: `VMContextWithoutEnv` and `VMContextWithEnv`. To clarify things, wouldn't it be possible to write this union as follows:

```rust
pub enum Either<L, R> {
    Left(L),
    Right(R),
}

pub union VMFunctionContextParameter {
    pub vmctx: *mut VMContext,
    pub host_env: Either<*mut VMContextWithEnv, *mut VMContextWithoutEnv>,
}
```

`mem::size_of::<VMFunctionContextParameter>` would be around 16 bytes, which is quite “heavy”. Without `Either`, it's 8 bytes.

If we don't really care about the size of this pointer, I would suggest to use `Either` or a similar enum, to be super explicity about what kind of pointer `host_env` can hold.

Reply:
My understanding is that there are 3 states that can exist on this single pointer:

1. The pointer is used in the context of a compiled Wasm function. The pointer is a `VMContext`.
2. The pointer is used in the context of a host function with an environment.  The pointer is a pointer to user data.
3. The pointer is used in the context of a host function without an environment.  The pointer is never used and we usually set it to null.

I don't think we need to represent the difference between 2 and 3 really, but I think the difference between 1 and 2 can be quite important.  Please let me know if my mental model of this is not the same as yours!

Reply:
Your mental model matches mine. I don't know if that's a good news though 🤔.

😄